### ISSUE TITLE
The ability to specify a path to a specific python version, rather than relying on python3

### ISSUE THREAD

**Issue by z0xyz** [williamboman/mason.nvim#798]
### I've searched open issues for similar requests

Yes

### I've manually reviewed logs to find potential errors

Yes

### I've recently downloaded the latest plugin version of mason.nvim

- [X] Yes

### Problem description

> what the issue is

When I open java files, a warning message at the status bar states that `Client 1 quit with exit code 1 and signal 0`, then the language server simply doesn't work.

> why you think it's an issue with mason.nvim.

issuing `LspLong` states at the very end that

```
[INFO][2022-12-25 14:00:58] .../vim/lsp/rpc.lua:661	"Starting RPC client"	{  args = {},  cmd = "jdtls",  extra = {    cwd = "/home/sinful"  }}
[TRACE][2022-12-25 14:00:58] .../lua/vim/lsp.lua:1276	"LSP[jdtls]"	"initialize_params"	{  capabilities = {    callHierarchy = {      dynamicRegistration = false    },    textDocument = {      codeAction = {        codeActionLiteralSupport = {          codeActionKind = {            valueSet = { "", "Empty", "QuickFix", "Refactor", "RefactorExtract", "RefactorInline", "RefactorRewrite", "Source", "SourceOrganizeImports", "quickfix", "refactor", "refactor.extract", "refactor.inline", "refactor.rewrite", "source", "source.organizeImports" }          }        },        dataSupport = true,        dynamicRegistration = false,        isPreferredSupport = true,        resolveSupport = {          properties = { "edit" }        }      },      completion = {        completionItem = {          commitCharactersSupport = false,          deprecatedSupport = false,          documentationFormat = { "markdown", "plaintext" },          preselectSupport = false,          snippetSupport = false        },        completionItemKind = {          valueSet = { 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25 }        },        contextSupport = false,        dynamicRegistration = false      },      declaration = {        linkSupport = true      },      definition = {        linkSupport = true      },      documentHighlight = {        dynamicRegistration = false      },      documentSymbol = {        dynamicRegistration = false,        hierarchicalDocumentSymbolSupport = true,        symbolKind = {          valueSet = { 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26 }        }      },      hover = {        contentFormat = { "markdown", "plaintext" },        dynamicRegistration = false      },      implementation = {        linkSupport = true      },      publishDiagnostics = {        relatedInformation = true,        tagSupport = {          valueSet = { 1, 2 }        }      },      references = {        dynamicRegistration = false      },      rename = {        dynamicRegistration = false,        prepareSupport = true      },      signatureHelp = {        dynamicRegistration = false,        signatureInformation = {          activeParameterSupport = true,          documentationFormat = { "markdown", "plaintext" },          parameterInformation = {            labelOffsetSupport = true          }        }      },      synchronization = {        didSave = true,        dynamicRegistration = false,        willSave = false,        willSaveWaitUntil = false      },      typeDefinition = {        linkSupport = true      }    },    window = {      showDocument = {        support = false      },      showMessage = {        messageActionItem = {          additionalPropertiesSupport = false        }      },      workDoneProgress = true    },    workspace = {      applyEdit = true,      configuration = true,      symbol = {        dynamicRegistration = false,        hierarchicalWorkspaceSymbolSupport = true,        symbolKind = {          valueSet = { 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26 }        }      },      workspaceEdit = {        resourceOperations = { "rename", "create", "delete" }      },      workspaceFolders = true    }  },  clientInfo = {    name = "Neovim",    version = "0.8.1"  },  initializationOptions = {    jvm_args = {},    workspace = "/home/runner/.cache/jdtls/workspace"  },  processId = 691275,  rootPath = vim.NIL,  rootUri = vim.NIL,  trace = "off",  workspaceFolders = vim.NIL}
[DEBUG][2022-12-25 14:00:58] .../vim/lsp/rpc.lua:285	"rpc.send"	{  id = 1,  jsonrpc = "2.0",  method = "initialize",  params = {    capabilities = {      callHierarchy = {        dynamicRegistration = false      },      textDocument = {        codeAction = {          codeActionLiteralSupport = {            codeActionKind = {              valueSet = { "", "Empty", "QuickFix", "Refactor", "RefactorExtract", "RefactorInline", "RefactorRewrite", "Source", "SourceOrganizeImports", "quickfix", "refactor", "refactor.extract", "refactor.inline", "refactor.rewrite", "source", "source.organizeImports" }            }          },          dataSupport = true,          dynamicRegistration = false,          isPreferredSupport = true,          resolveSupport = {            properties = { "edit" }          }        },        completion = {          completionItem = {            commitCharactersSupport = false,            deprecatedSupport = false,            documentationFormat = { "markdown", "plaintext" },            preselectSupport = false,            snippetSupport = false          },          completionItemKind = {            valueSet = { 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25 }          },          contextSupport = false,          dynamicRegistration = false        },        declaration = {          linkSupport = true        },        definition = {          linkSupport = true        },        documentHighlight = {          dynamicRegistration = false        },        documentSymbol = {          dynamicRegistration = false,          hierarchicalDocumentSymbolSupport = true,          symbolKind = {            valueSet = { 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26 }          }        },        hover = {          contentFormat = { "markdown", "plaintext" },          dynamicRegistration = false        },        implementation = {          linkSupport = true        },        publishDiagnostics = {          relatedInformation = true,          tagSupport = {            valueSet = { 1, 2 }          }        },        references = {          dynamicRegistration = false        },        rename = {          dynamicRegistration = false,          prepareSupport = true        },        signatureHelp = {          dynamicRegistration = false,          signatureInformation = {            activeParameterSupport = true,            documentationFormat = { "markdown", "plaintext" },            parameterInformation = {              labelOffsetSupport = true            }          }        },        synchronization = {          didSave = true,          dynamicRegistration = false,          willSave = false,          willSaveWaitUntil = false        },        typeDefinition = {          linkSupport = true        }      },      window = {        showDocument = {          support = false        },        showMessage = {          messageActionItem = {            additionalPropertiesSupport = false          }        },        workDoneProgress = true      },      workspace = {        applyEdit = true,        configuration = true,        symbol = {          dynamicRegistration = false,          hierarchicalWorkspaceSymbolSupport = true,          symbolKind = {            valueSet = { 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26 }          }        },        workspaceEdit = {          resourceOperations = { "rename", "create", "delete" }        },        workspaceFolders = true      }    },    clientInfo = {      name = "Neovim",      version = "0.8.1"    },    initializationOptions = {      jvm_args = {},      workspace = "/home/runner/.cache/jdtls/workspace"    },    processId = 691275,    rootPath = vim.NIL,    rootUri = vim.NIL,    trace = "off",    workspaceFolders = vim.NIL  }}
[ERROR][2022-12-25 14:00:59] .../vim/lsp/rpc.lua:733	"rpc"	"jdtls"	"stderr"	"Traceback (most recent call last):\n  File \"/home/sinful/.local/share/nvim/mason/bin/jdtls\", line 17, in <module>\n    jdtls.main(sys.argv[1:])\n  File \"/home/sinful/.local/share/nvim/mason/packages/jdtls/bin/jdtls.py\", line 73, in main\n    parser.add_argument(\"--validate-java-version\", default=True, action=argparse.BooleanOptionalAction)\nAttributeError: module 'argparse' has no attribute 'BooleanOptionalAction'\n"
```
Which turned up to be a problem of the project relying on a newer version of python's standard library `argparse` i..e python 3.9

### Expected behavior

The plugin giving the option to at least specify python version from its configuration, or backward compatibility to plain python3 versions

### Steps to reproduce

1. Install jdtls server
2. Make python3 points to a version < python3.9
3. Open a java file.

### Affected packages

jdtls

### Neovim version (>= 0.7)

 $ nvim --version
```
NVIM v0.8.1
Build type: Release
LuaJIT 2.1.0-beta3
Compiled by runner@fv-az178-366

Features: +acl +iconv +tui
See ":help feature-compile"

   system vimrc file: "$VIM/sysinit.vim"
  fall-back for $VIM: "/share/nvim"

Run :checkhealth for more info
```

### Operating system/version

Linux unixhost 5.15.0-56-generic #62~20.04.1-Ubuntu SMP Tue Nov 22 21:24:20 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux

### Healthcheck

```Text
mason: require("mason.health").check()
========================================================================
## mason.nvim report
  - OK: neovim version >= 0.7.0
  - WARNING: **Go**: unsupported version `go version go1.13.8 linux/amd64`. Go version must be >= 1.17.
  - WARNING: **cargo**: unsupported version `cargo 1.59.0 (49d8809dc 2022-02-10)`. Some cargo installations require Rust >= 1.60.0.
  - WARNING: **luarocks**: not available
  - OK: **Ruby**: `ruby 2.7.0p0 (2019-12-25 revision 647ee6f091) [x86_64-linux-gnu]`
  - OK: **RubyGem**: `3.1.2`
  - WARNING: **Composer**: not available
  - WARNING: **PHP**: not available
  - OK: **npm**: `8.5.2`
  - OK: **node**: `v17.7.2`
  - OK: **python3**: `Python 3.8.10`
  - OK: **pip3**: `pip 22.3.1 from /home/sinful/.local/lib/python3.8/site-packages/pip (python 3.8)`
  - OK: **javac**: `javac 17.0.5`
  - OK: **java**: `openjdk version "17.0.5" 2022-10-18`
  - WARNING: **julia**: not available
  - OK: **wget**: `GNU Wget 1.20.3 built on linux-gnu.`
  - OK: **curl**: `curl 7.68.0 (x86_64-pc-linux-gnu) libcurl/7.68.0 OpenSSL/1.1.1f zlib/1.2.11 brotli/1.0.7 libidn2/2.2.0 libpsl/0.21.0 (+libidn2/2.2.0) libssh/0.9.3/openssl/zlib nghttp2/1.40.0 librtmp/2.3`
  - OK: **gzip**: `gzip 1.10`
  - OK: **tar**: `tar (GNU tar) 1.30`
  - WARNING: **pwsh**: not available
  - OK: **bash**: `GNU bash, version 5.0.17(1)-release (x86_64-pc-linux-gnu)`
  - OK: **sh**: `Ok`
  - OK: GitHub API rate limit. Used: 0. Remaining: 60. Limit: 60. Reset: Sun 25 Dec 2022 03:09:00 PM EET.
```
```


### Screenshots or recordings

_No response_

**Comment by williamboman** [comment]
Hello! So this is a bit of a complicated topic with, at least to me, somewhat fuzzy responsibilities in terms of what Mason should and shouldn't do. Currently it resides outside of Mason's responsibilities.

The `jdtls` executable is a shell script with a `#!/usr/bin/env python3` shebang. This effectively means that when you launch the `jdtls` server, it'll simply use the first `python3` in your `PATH` (which likely ends up being some python version in your activated virtual env). The way Neovim's LSP client currently works is that it simply inherits the `PATH` of the Neovim instance, which (although depending on how you run Neovim) will inherit `PATH` from your shell. This means you'll have to ensure that when you launch Neovim you're in a "clean" environment, or alternatively amend the `PATH` somewhere in your Neovim init script.

However, I definitely think Mason could help manage runtimes by installing a pinned version of them (e.g. `python@3.10` and `node@18`) and ensuring that these are used when executing executables installed via Mason. This could for example be done through [shimming](https://en.wikipedia.org/wiki/Shim_(computing)) (more complex) or modifying the `PATH` before Neovim's LSP client launches (less complex but constrained to LSP use-case only).

**Comment by williamboman** [comment]
FWIW Mason.nvim does respect the `vim.g.python3_host_prog` variable when Mason itself invokes python (i.e. when installing python packages).

**Comment by z0xyz** [comment]
> Currently it resides outside of Mason's responsibilities.

It's just that Mason is not backward compatible with a pretty recent python version i.e. `python3.8`. The problem lies upfront at Mason itself due to its reliance on a recent `argparse` syntax, not with jdtls in this case (at least for now).

> This means you'll have to ensure that when you launch Neovim you're in a "clean" environment, or alternatively amend the PATH somewhere in your Neovim init script.

A bit messy. The `PATH` should be confined to the shell's init script.

> This could for example be done through [shimming](https://en.wikipedia.org/wiki/Shim_(computing)

How would I go about that. Learn lua, then what. Some overall guidance would be really useful.

> Mason does respect the vim.g.python3_host_prog variable when Mason itself invokes python (i.e. when installing python packages).

I've set this variable at my config file, and during runtime (started the LSP server afterwards), and didn't find any change. 

>  modifying the PATH before Neovim's LSP client launches (less complex but constrained to the LSP use-case only).

I did update python alternatives, but it breaks other linux tools. switching back and forth is a hassle.

**Comment by williamboman** [comment]
> It's just that Mason is not backward compatible with a pretty recent python version i.e. python3.8. The problem lies upfront at Mason itself due to its reliance on a recent argparse syntax, not with jdtls in this case (at least for now).

Not Mason, see:
- https://github.com/eclipse/eclipse.jdt.ls/blob/master/org.eclipse.jdt.ls.product/scripts/jdtls
- https://github.com/eclipse/eclipse.jdt.ls/blob/master/org.eclipse.jdt.ls.product/scripts/jdtls.py

> How would I go about that. Learn lua, then what. Some overall guidance would be really useful.

Oh it was just me thinking out loud what could be done in terms of future development.

> I've set this variable at my config file, and during runtime (started the LSP server afterwards), and didn't find any change. Respecting or prioritizing this variable over that python version obtained via PATH seems to be the most viable option.

Going back to my earlier point, Mason is not involved in managing which runtime is used when invoking a package executable. That's delegated to each user's PATH. What I meant was that the python3_host_prog variable is respected during installation of Python packages.

For jdtls specifically you can always avoid the `jdtls` python executable and instead run the jar file immediately (per their docs).

**Comment by z0xyz** [comment]
> Not Mason, see:

Oh, Sorry. Didn't mind the intermediary jdtls server subdirectory.

> For jdtls specifically you can always avoid the jdtls python executable and instead run the jar file immediately (per their docs).

There's a issue opened concerning this very problem opened at the other repo, and the dev is open for a PR. I'll continue from there.

Sorry for that. I did previously look at jdtls's opened issues (as I suspected both Mason, and jdtls), but only now did i find the issue out there. Thanks for the suggestions & hints.
